### ISSUE TITLE
Tutorial Code Hangs

### ISSUE THREAD

**Issue by mabdullahsoyturk** [nict-wisdom/rannc#6]
After I did the installation according to the [installation page](https://nict-wisdom.github.io/rannc/installation.html) in the documentation, I tried to run the [tutorial script](https://nict-wisdom.github.io/rannc/tutorial.html) but it hangs for some reason. 

I am trying it on 2 A100 GPUs. I use the following software:

- CUDA version 10.2
- NCCL version 2.11.4
- Python 3.8.12
- MPI version openmpi-4.0.5

The command that I used is exactly the same with the tutorial: 

```bash
mpirun -np 2 python tutorial.py 64 512 10
```

When I run this command this is the output:

[2022-01-20 13:19:42.900] [RaNNCProcess] [info] RaNNC started on rank 0 (g002)
[2022-01-20 13:19:42.900] [RaNNCProcess] [info] RaNNC started on rank 1 (g002)
[2022-01-20 13:19:42.901] [RaNNCProcess] [info] CUDA device assignments:
[2022-01-20 13:19:42.901] [RaNNCProcess] [info]  Worker 0: device0@g002
[2022-01-20 13:19:42.901] [RaNNCProcess] [info]  Worker 1: device1@g002
#Parameters=2626560

It hangs after this point.

**Comment by nict-wisdom** [comment]
Thank you for your report!

I created a new environment and installed RaNNC, but couldn't reproduce the problem.

RaNNC calls MPI_Allreduce using MPI soon after the script shows "#Parameters=2626560".
So I guess it causes the program to hang.
To test the call, can you add `gather_inputs=False` as an argument of pyrannc.RaNNCModule()
```
model = pyrannc.RaNNCModule(model, opt, gather_inputs=False)
```

RaNNC skips the call of MPI_Allreduce when this option is given.
We can see whether the MPI call causes the problem or not.

**Comment by mabdullahsoyturk** [comment]
It solved the hanging issue but now I am getting a segmentation fault. 

Here is the output: 

```
[2022-01-25 10:51:35.089] [RaNNCProcess] [info] RaNNC started on rank 0 (g001)
[2022-01-25 10:51:35.097] [RaNNCProcess] [info] RaNNC started on rank 1 (g001)
[2022-01-25 10:51:35.109] [RaNNCProcess] [info] CUDA device assignments:
[2022-01-25 10:51:35.109] [RaNNCProcess] [info]  Worker 0: device0@g001
[2022-01-25 10:51:35.109] [RaNNCProcess] [info]  Worker 1: device1@g001
#Parameters=2626560
[2022-01-25 10:51:45.230] [RaNNCModule] [info] Tracing model ...
[2022-01-25 10:51:46.105] [RaNNCModule] [info] Converting torch model to IR ...
[2022-01-25 10:51:47.149] [RaNNCModule] [info] Running profiler ...
[2022-01-25 10:51:47.302] [RaNNCModule] [info] Profiling finished
[2022-01-25 10:51:47.303] [RaNNCModule] [info] Assuming batch size: 64
[2022-01-25 10:51:47.305] [Decomposer] [info] Decomposer: ml_part
[2022-01-25 10:51:47.306] [Decomposer] [info] Available device memory: 30680756224
[2022-01-25 10:51:47.306] [Decomposer] [info] Starting model partitioning ... (this may take a very long time)
[2022-01-25 10:51:48.812] [DPStaging] [info] Estimated profiles of subgraphs: batch_size=64 np=2 pipeline=1 use_amp=0 zero=0
  graph=MERGE_0_9 repl=2 fwd_time=5103 bwd_time=51389 ar_time=978 in_size=65536 out_size=65536 fp32param_size=10506240 fp16param_size=0 total_mem=53645312 (fwd+bwd=32436224 opt=21012480 comm=196608)
[2022-01-25 10:51:48.813] [Decomposer] [info]  Assigned subgraph MERGE_0_9 to rank[1,0]
[2022-01-25 10:51:48.814] [RaNNCModule] [info] Routes verification passed.
[2022-01-25 10:51:48.936] [ParamStorage] [info] Synchronizing parameters ...
[2022-01-25 10:51:49.014] [RaNNCModule] [info] RaNNCModule is ready. (rank0)
[2022-01-25 10:51:49.014] [RaNNCModule] [info] RaNNCModule is ready. (rank1)
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 21197 on node g001 exited on signal 11 (Segmentation fault).
```

**Comment by nict-wisdom** [comment]
I released a new version (v0.7.3) to solve the problems.
Can you try it? (See [Installation](https://nict-wisdom.github.io/rannc/installation.html))

I think the MPI installation in your environment doesn't work as I expected.
We found that `gather_inputs=False` causes a segmentation fault when the MPI installation does not support CUDA.
In addition, the output you showed suggested that allreduce using MPI hangs.

In the new version, the MPI calls are replaced with ones using NCCL.
I hope this solves the problems.

**Comment by mabdullahsoyturk** [comment]
Thanks for the quick answer. It works now but I am pretty sure that the mpi installation on the cluster supports CUDA though.
